lti_consumer/lti_xblock.py

```diff
@@ -1111,13 +1091,22 @@ def lti_launch_handler(self, request, suffix=''):  # pylint: disable=unused-argu
         if self.ask_to_send_email and real_user_data['user_email']:
             email = real_user_data['user_email']
 
-        lti_consumer.set_user_data(
-            self.user_id,
-            self.role,
-            result_sourcedid=self.lis_result_sourcedid,
-            person_sourcedid=username,
-            person_contact_email_primary=email
-        )
+        # Occassionally, users try to do an LTI launch while they are unauthenticated. It is not known why this occurs.
+        # Sometimes, it is due to a web crawlers; other times, it is due to actual users of the platform. Regardless,
+        # return a 400 response with an appropriate error template.
+        try:
```

Comment:
this try feels a little weird to me because what can throw the error is self.user_id, not set_user_data which cannot throw any errors

also why doesn't extract_real_user_data above also fail?

maybe a little more restructuring to make this function well ordered:
first get the consumer
then get the various user fields inside a try block indicating "there might not be a user"
then set the user data off those fields which can't fail

Reply:
I figured that every call to the user service should care whether the user is authenticated, so I added that logic.

Reply:
thanks, this is clearer